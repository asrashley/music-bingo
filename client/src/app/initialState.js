import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
