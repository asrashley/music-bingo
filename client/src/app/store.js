import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
