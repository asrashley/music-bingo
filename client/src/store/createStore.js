import { configureStore } from '@reduxjs/toolkit';
import { createLogger } from 'redux-logger';
import { createRouterMiddleware, createRouterReducer } from '@lagunovsky/redux-react-router';
import log from 'loglevel';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import routesReducer from '../routes/routesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';
const routerMiddleware = createRouterMiddleware(history);

export function createStore(preloadedState, logging) {
  const middleware = (getDefaultMiddleware) => {
    const mw = getDefaultMiddleware()
      .prepend(routerMiddleware)
      .concat(usersMiddleware)
      .concat(messagesMiddleware);
    if (logging === undefined) {
      logging = import.meta.env.DEV;
    }
    if (!logging) {
      return mw;
    }
    const logger = {
      debug: log.debug,
      info: log.info,
      warn: log.warn,
      error: log.error,
      log: log.debug,
    };
    const loggerMiddleware = createLogger({
      level: 'debug',
      logger
    });
    return mw.concat(loggerMiddleware);
  };
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      routes: routesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: createRouterReducer(history),
      user: userReducer,
    },
    preloadedState
  });
}
