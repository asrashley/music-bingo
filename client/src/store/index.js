export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
