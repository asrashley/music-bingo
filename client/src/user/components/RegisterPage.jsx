import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { reverse } from 'named-urls';
import log from 'loglevel';

import { RegisterForm } from './RegisterForm';
import { checkUser, registerUser } from '../userSlice';
import routes from '../../routes';

import '../styles/user.scss';

class RegisterPage extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func,
    history: PropTypes.object,
  };


  handleSubmit = ({ email, password, username }) => {
    const { dispatch } = this.props;
    return dispatch(registerUser({ email, password, username })).then(this.submitResponse);
  };

  submitResponse = (result) => {
    const { history } = this.props;
    if (!result || !result.payload) {
      return [{
        type: "validate",
        message: 'Unknown error',
        name: 'email',
      }];
    }
    const { payload } = result;
    const { success, error } = payload;
    if (success === true) {
      history.push(reverse(`${routes.index}`));
      return true;
    }

    const errs = [];
    for (let name in error) {
      if (error[name]) {
        errs.push({
          type: "validate",
          message: error[name],
          name,
        });
      }
    }
    return(errs);
  };

  onCancel = () => {
    const { history } = this.props;
    log.debug('Registration cancelled');
    history.push(reverse(`${routes.login}`));
  };

  checkUser = ({ email, username }) => {
    const { dispatch } = this.props;
    return dispatch(checkUser({ email, username }))
      .then((result) => {
        if (!result) {
          return ({ error: 'Unknown error' });
        }
        return result.payload;
      })
      .catch(err => {
        if (typeof (err) === 'object') {
          const { status, statusText } = err;
          if (status === 500) {
            return ({ error: "There is a problem with the server. Please try again later" });
          }
          if (statusText !== undefined) {
            return ({ error: statusText });
          }
          return ({ error: 'Unknown error' });
        }
        return ({ error: `${err}` });
      });
  }

  render() {
    return (
      <RegisterForm {...this.props} onSubmit={this.handleSubmit} onCancel={this.onCancel}
        checkUser={this.checkUser}
      />
    );
  }
};

RegisterPage = connect()(RegisterPage);

export {
  RegisterPage
};
