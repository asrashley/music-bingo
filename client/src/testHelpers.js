import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import log from 'loglevel';
import { screen, fireEvent } from '@testing-library/react';
import waitForExpect from 'wait-for-expect';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import { history } from './store/history';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ConnectedRouter history={history}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ConnectedRouter>
    </Provider>);
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
};

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    if (modFn !== undefined) {
      return jsonResponse(modFn(url, data['default']));
    }
    return jsonResponse(data['default']);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = (url, opts) => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async (url, opts) => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    return jsonResponse(data['default']);
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    const userIsValid = userDatabase.some(item => (
      item.username === username && item.password === password));
    log.debug(`loginUser: username="${username}" password="${password}" valid=${userIsValid}`);
    if (!userIsValid) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    loggedIn = true;
    return jsonResponse(data['default']);
  };
  const logoutUser = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  fetchMock.config.fallbackToNetwork = false;
  fetchMock.config.warnOnFallback = true;
  log.debug(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields) {
  fields.forEach(({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    fireEvent.input(screen.getByLabelText(label, { exact }), {
      target: {
        value
      }
    });
  });
  await waitForExpect(() => {
    const last = fields[fields.length - 1];
    const { value } = document.querySelector(`input[name="${last.label}"`);
    expect(value).toBe(fields[fields.length - 1].value);
  });
}

