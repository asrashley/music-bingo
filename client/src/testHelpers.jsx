import { render, queries, buildQueries, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import { ReduxRouter } from '@lagunovsky/redux-react-router';
import log from 'loglevel';
import { Readable } from "stream";
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import fetchMockHandlerTest from 'fetch-mock/esm/client';
import jestify from 'fetch-mock-jest/jestify';

import userData from './fixtures/user.json';

export const fetchMock = jestify(fetchMockHandlerTest);

/* custom query that looks for a "data-last-update" attribute in the container and compares that to
  the value provided to the query. It can be used to check if a component has been updated or
  to wait for it to re-render
  */
const queryAllByLastUpdate = (container, lastUpdate, options = {}) => {
  const { comparison } = {
    comparison: 'equals',
    ...options
  };
  log.trace(`queryAllByLastUpdate lastUpdate=${lastUpdate} comparison=${comparison} container="${container.nodeName}.${container.className}" ${container.id}`);
  return Array.from(container.querySelectorAll('[data-last-update]'))
    .filter((elt) => {
      const update = parseInt(elt.dataset.lastUpdate, 10);
      log.trace(`${elt.nodeName}.${elt.className} lastUpdate = ${update}`);
      if (isNaN(update)) {
        return false;
      }
      switch (comparison) {
        case 'equals':
          return update === lastUpdate;
        case 'greaterThan':
          return update > lastUpdate;
        case 'greaterThanOrEquals':
          return update >= lastUpdate;
        case 'lessThan':
          return update < lastUpdate;
        case 'lessThanOrEquals':
          return update <= lastUpdate;
        default:
          return true;
      }
    });
};
export const [queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate] = buildQueries(
  queryAllByLastUpdate,
  (container, selector) => `Found multiple elements from ${container} with last update selector: ${selector}`,
  (container, selector, opts = {}) => {
    const { comparison = 'equals' } = opts;
    return (`Unable to find an element from ${container.nodeName}.${container.className} with last update ${comparison} ${selector}`);
  }
);
const lastUpdatedQueries = { queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate };

/* custom query that uses a CSS selector to find elements */
const queryAllBySelector = (container, selector) => {
  log.trace(`queryAllBySelector ${selector}`);
  return Array.from(container.querySelectorAll(selector));
};
const [queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector] = buildQueries(
  queryAllBySelector,
  (container, selector) => `Found multiple elements from ${container} with selector: ${selector}`,
  (container, selector) => `Unable to find an element from ${container} with selector: ${selector}`,
);
const bySelectorQueries = { queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector };

const routerSelector = (state) => state.router

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    history = createMemoryHistory(),
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ReduxRouter history={history} routerSelector={routerSelector}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ReduxRouter>
    </Provider>);
  }
  Wrapper.propTypes = {
    children: PropTypes.node,
  };
  const events = userEvent.setup({
    delay: 10,
  });
  return {
    events,
    store,
    ...render(ui, {
      wrapper: Wrapper,
      queries: {
        ...queries,
        ...lastUpdatedQueries,
        ...bySelectorQueries
      },
      ...renderOptions
    })
  };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Cache-Control': 'max-age = 0, no_cache, no_store, must_revalidate',
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
}

export class MockResponse extends Response {
  constructor(...args) {
    if (args[0] instanceof ReadableStream) {
      args[0] = Readable.from(args[0]);
    }
    super(...args);
  }
}

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    let responsePayload = data['default'];
    if (modFn !== undefined) {
      log.debug(`modify response for ${url}`);
      responsePayload = modFn(url, responsePayload);
      log.debug(`modify response for ${url} done`);
    }
    return jsonResponse(responsePayload);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = () => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async () => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    return jsonResponse({
      ...userData,
      refreshToken,
      accessToken: getAccessToken(),
    });
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    let index = -1;
    let dbEntry;
    userDatabase.forEach((item, i) => {
      if (item.username === username && item.password === password) {
        dbEntry = item;
        index = i;
      }
    });
    log.debug(`loginUser: username="${username}" password="${password}" index=${index}`);
    if (!dbEntry) {
      return 401;
    }
    const user = {
      ...userData,
      pk: index + 2,
      ...dbEntry,
      refreshToken,
      accessToken: getAccessToken()
    };
    delete user.password;
    loggedIn = true;
    return jsonResponse(user);
  };
  const logoutUser = () => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  Object.assign(fetchMock.config, {
    fallbackToNetwork: false,
    warnOnFallback: true,
    Response: MockResponse,
  });
  log.trace(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    logout: () => loggedIn = false,
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields, events) {
  await Promise.all(fields.map(async ({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    const elt = await screen.findByLabelText(label, { exact });
    if (events) {
      //await events.click(elt);
      await waitFor(() => events.clear(elt));
      /*
            await waitFor(() => events.clear(elt));
        await waitFor(() => events.type(elt, value));
      }*/
    }
    await waitFor(() => fireEvent.input(elt, {
      target: {
        value
      }
    }));
  }));
  const { label, exact, value } = fields[fields.length - 1];
  const elt = await screen.findByLabelText(label, { exact });
  await waitFor(() => {
    expect(elt.value).toBe(value);
  });
}

