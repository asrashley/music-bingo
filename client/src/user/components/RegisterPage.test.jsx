import React from 'react';
import { fireEvent, screen } from '@testing-library/react';
import log from 'loglevel';
import fetchMock from "fetch-mock-jest";
import waitForExpect from 'wait-for-expect';

import { renderWithProviders, installFetchMocks, setFormFields } from '../../testHelpers';
import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';

import { RegisterPage } from './RegisterPage';

import * as user from '../../fixtures/userState.json';

function setFormValues({ email, username, password }) {
  setFormFields([
    {
      label: "Email address",
      value: email,
      exact: false
    },
    {
      label: "Username",
      value: username,
      exact: false
    },
    {
      label: /^Password/,
      value: password
    },
    {
      label: /^Confirm Password/,
      value: password
    },
  ]);
}

describe('RegisterPage component', () => {
  let store;
  let apiMock;

  beforeEach(() => {
    store = createStore(initialState);
    apiMock = installFetchMocks(fetchMock, { loggedIn: false });
  });

  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
  });

  it('renders from initial state', () => {
    const { dispatch } = store;
    const history = {
      push: jest.fn()
    };
    const props = {
      dispatch,
      history
    };
    const result = renderWithProviders(<RegisterPage {...props} />, { store });
    expect(result.asFragment()).toMatchSnapshot();
  });

  it('redirects page when register is successful', async () => {
    const { dispatch } = store;
    const history = {
      push: jest.fn()
    };
    const props = {
      dispatch,
      history
    };
    const expected = {
      email: 'successful@unit.test',
      username: 'newuser',
      password: '!secret!',
      rememberme: false
    };
    const addUserApi = jest.fn((url, opts) => {
      const { email, username, password } = JSON.parse(opts.body);
      expect(email).toBe(expected.email);
      expect(username).toBe(expected.username);
      expect(password).toBe(expected.password);
      return apiMock.jsonResponse({
        'message': 'Successfully registered',
        'success': true,
        'user': {
          ...user,
          email,
          username,
          groups: ["users"]
        },
        'accessToken': apiMock.getAccessToken(),
        'refreshToken': apiMock.getRefreshToken(),
      });
    });
    fetchMock.put('/api/user', addUserApi);
    //log.setLevel('debug');
    renderWithProviders(<RegisterPage {...props} />);
    //screen.debug();
    setFormValues(expected);
    fireEvent.submit(await screen.findByText('Register'));
    await waitForExpect(() => {
      expect(addUserApi).toHaveBeenCalledTimes(1);
    });
    await waitForExpect(() => {
      expect(history.push).toHaveBeenCalledTimes(1);
    });
  });

  it.each(['username', 'email'])('shows error message when %s is already in use', async (field) => {
    const { dispatch } = store;
    const history = {
      push: jest.fn()
    };
    const props = {
      dispatch,
      history
    };
    const expected = {
      email: 'a.user@example.tld',
      username: 'user',
      password: 'wrongpassword',
      rememberme: false
    };
    if (field === 'username') {
      apiMock.addUser({
        email: 'a.different@email.net',
        username: expected.username
      });
    } else {
      apiMock.addUser({
        email: expected.email,
        username: 'anotheruser'
      });
    }
    //log.setLevel('debug');
    renderWithProviders(<RegisterPage {...props} />);
    setFormValues(expected);
    fireEvent.submit(await screen.findByText('Register'));
    if (field === 'email') {
      await screen.findByText('That email address is already registered');
    } else {
      await screen.findByText('That username is already taken');
    }
  });

  it('shows error message when there is a server fault', async () => {
    const { dispatch } = store;
    const history = {
      push: jest.fn()
    };
    const props = {
      dispatch,
      history
    };
    const expected = {
      email: 'a.user@example.tld',
      username: 'user',
      password: 'mysecret',
      rememberme: false
    };
    //log.setLevel('debug');
    renderWithProviders(<RegisterPage {...props} />);
    setFormValues(expected);
    apiMock.setServerStatus(500);
    fetchMock.put('/api/user', (url, opts) => {
      return 500;
    });
    fireEvent.submit(await screen.findByText('Register'));
    await screen.findAllByText("500: Internal Server Error");
  });

  it('shows error message from server', async () => {
    const { dispatch } = store;
    const history = {
      push: jest.fn()
    };
    const props = {
      dispatch,
      history
    };
    const expected = {
      email: 'successful@unit.test',
      username: 'newuser',
      password: '!secret!',
      rememberme: false
    };
    const addUserApi = jest.fn((url, opts) => {
      const { email, username, password } = JSON.parse(opts.body);
      expect(email).toBe(expected.email);
      expect(username).toBe(expected.username);
      expect(password).toBe(expected.password);
      return apiMock.jsonResponse({
        'success': false,
        'error': {
          'username': `Username ${username} is already taken`
        },
        user: {
          email,
          username
        }
      });
    });
    fetchMock.put('/api/user', addUserApi);
    //log.setLevel('debug');
    renderWithProviders(<RegisterPage {...props} />);
    //screen.debug();
    setFormValues(expected);
    fireEvent.submit(await screen.findByText('Register'));
    await waitForExpect(() => {
      expect(addUserApi).toHaveBeenCalledTimes(1);
    });
    await screen.findByText(`Username ${expected.username} is already taken`);
  });

});
