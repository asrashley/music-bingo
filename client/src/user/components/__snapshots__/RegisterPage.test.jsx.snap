// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RegisterPage component > renders from initial state 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      id="message-panel"
    />
    <form
      class="register-form"
    >
      <div
        class="form-group"
      >
        <label
          for="field-username"
        >
          Username
        </label>
        <input
          class="form-control   "
          id="field-username"
          name="username"
          placeholder="Username"
          type="text"
        />
        <small
          class="form-text text-muted"
        >
          This is the name you will use to log in and the name that will be visible to other players.
        </small>
      </div>
      <div
        class="form-group"
      >
        <label
          for="field-email"
        >
          Email address
        </label>
        <input
          class="form-control   "
          id="field-email"
          name="email"
          placeholder="Email address"
          type="text"
        />
        <small
          class="form-text text-muted"
        >
          We'll never share your email with anyone else.
        </small>
      </div>
      <div
        class="form-group"
      >
        <label
          for="field-password"
        >
          Password
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control password  "
          id="field-password"
          name="password"
          placeholder="Password"
          type="password"
        />
        <small
          class="form-text text-muted"
        >
          The password needs to be at least 5 characters in length.
        </small>
      </div>
      <div
        class="form-group"
      >
        <label
          for="field-confirmPassword"
        >
          Confirm Password
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control password  "
          id="field-confirmPassword"
          name="confirmPassword"
          placeholder="Confirm Password"
          type="password"
        />
        <small
          class="form-text text-muted"
        >
          Please enter the password a second time to make sure you have typed it correctly
        </small>
      </div>
      <div
        class="form-text text-muted"
      >
        * = a required field
      </div>
      <div
        class="form-group modal-footer"
      >
        <button
          class="btn btn-primary"
          disabled=""
          type="submit"
        >
          Register
        </button>
        <button
          class="btn btn-danger"
          name="cancel"
          type="cancel"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
